# Gradient Boosting - Hyperparameter Tuning

The previous notebook explored selected Gradient Boosting hyperparameters manually. Increasing tree depth produced the largest improvement in fraud detection performance, but deeper trees also showed diminishing returns and an increasing risk of overfitting.

This notebook extends the manual experiments by introducing systematic hyperparameter tuning.

The tuning process will:

- explore multiple Gradient Boosting hyperparameter configurations,
- use F1-score as the primary optimization metric,
- preserve the temporal structure of the transaction data during validation,
- keep the 2018 validation set separate from hyperparameter tuning,
- keep 2019 completely unseen for final model evaluation.

## Experimental Design

The available labeled transaction history is divided according to time:

- **2010–2017:** model development and hyperparameter tuning
- **2018:** validation and model comparison
- **2019:** final out-of-time test set

Hyperparameter tuning must use only information available within the 2010–2017 development period. The temporal order of transactions will be preserved during cross-validation to avoid training models on future observations and evaluating them on the past.

In [60]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.impute import SimpleImputer

from sklearn.pipeline import Pipeline

from sklearn.model_selection import RandomizedSearchCV

from sklearn.ensemble import GradientBoostingClassifier

from sklearn.metrics import (
    recall_score,
    precision_score,
    f1_score,
    confusion_matrix,
)

from finsight.model_evaluation import evaluate_model

from finsight.fraud_preprocessing import (
    encode_categorical_features,
    handle_missing_values,
)

from finsight.database import (
    connect_to_database,
    test_database_connection,
)

from finsight.fraud_data import (
    BASE_FEATURES,
    download_yearly_data,
)

In [2]:
from finsight.utils.notebook_notifications import enable_cell_notifications

enable_cell_notifications()

In [3]:
engine = connect_to_database()
test_database_connection(engine)

Connected!


## Development Data

Hyperparameter tuning is performed exclusively on data from 2010–2017.

The development dataset is loaded separately for each year to preserve temporal boundaries required for time-aware cross-validation. Non-fraud transactions are sampled proportionally to the original yearly transaction volumes, with a total sampling budget of 300,000 observations.

The previously selected base feature set is retained so that the tuning experiments isolate the effect of model hyperparameters rather than changes in feature representation.

In [4]:
features = BASE_FEATURES

yearly_data = download_yearly_data(
    engine,
    start_year=2010,
    end_year=2017,
    total_non_fraud_limit=300000,
    features=features
)

In [5]:
for year, data in yearly_data.items():
    print(
        year,
        data["X"].shape,
        data["y"].value_counts().to_dict()
    )

2010 (37148, 13) {False: 34575, True: 2573}
2011 (36048, 13) {False: 36011, True: 37}
2012 (37815, 13) {False: 36892, True: 923}
2013 (39124, 13) {False: 37787, True: 1337}
2014 (38803, 13) {False: 38139, True: 664}
2015 (40896, 13) {False: 38707, True: 2189}
2016 (41251, 13) {False: 38803, True: 2448}
2017 (39258, 13) {False: 39086, True: 172}


## Temporal Development Set

The yearly datasets are concatenated in chronological order from 2010 to 2017.

The feature matrix and target vector are kept separate, while an additional year vector is created to identify the temporal origin of each observation. The year information is used only for constructing temporal validation folds and is not included as a model feature.

In [6]:
X_development = pd.concat(
    [yearly_data[year]["X"] for year in range(2010, 2018)],
    ignore_index=True
)
X_development.shape

(310343, 13)

In [7]:
y_development = pd.concat(
    [yearly_data[year]["y"] for year in range(2010, 2018)],
    ignore_index=True
)

y_development.shape

(310343,)

In [8]:
development_years = []

for year in yearly_data:
    development_years.extend(
        [year] * len(yearly_data[year]["X"])
    )

development_years = pd.Series(development_years)

development_years.shape

(310343,)

In [9]:
development_years.value_counts().sort_index()

2010    37148
2011    36048
2012    37815
2013    39124
2014    38803
2015    40896
2016    41251
2017    39258
Name: count, dtype: int64

## Temporal Cross-Validation

In [10]:
train_mask = development_years.isin([2010, 2011, 2012])
train_indices = development_years[train_mask].index
train_indices

RangeIndex(start=0, stop=111011, step=1)

In [11]:
development_years[train_mask].unique()

array([2010, 2011, 2012])

In [12]:
val_mask = development_years.isin([2013])
val_indices = development_years[val_mask].index
val_indices

RangeIndex(start=111011, stop=150135, step=1)

In [13]:
development_years[val_mask].unique()

array([2013])

In [14]:
train_indices.intersection(val_indices)

RangeIndex(start=0, stop=0, step=1)

### Automated Temporal Folds

In [15]:
def create_temporal_folds(development_years, min_years=3):
    folds = []
    years = development_years.unique()

    for val_year in years[min_years:]:
        val_mask = (development_years == val_year)
        val_indices = development_years[val_mask].index

        train_mask = development_years < val_year
        train_indices = development_years[train_mask].index

        folds.append((train_indices, val_indices))
        
    return folds

In [16]:
folds = create_temporal_folds(development_years)
len(folds)

5

In [17]:
for train_indices, val_indices in folds:
    print(development_years[train_indices].unique(), development_years[val_indices].unique())

[2010 2011 2012] [2013]
[2010 2011 2012 2013] [2014]
[2010 2011 2012 2013 2014] [2015]
[2010 2011 2012 2013 2014 2015] [2016]
[2010 2011 2012 2013 2014 2015 2016] [2017]


### Expanding Window Validation

To preserve the temporal structure of the transaction data, model configurations are evaluated using expanding-window cross-validation rather than random cross-validation.

Each fold trains the model using all available historical years and validates it on the immediately following year:

- Fold 1: 2010–2012 → 2013
- Fold 2: 2010–2013 → 2014
- Fold 3: 2010–2014 → 2015
- Fold 4: 2010–2015 → 2016
- Fold 5: 2010–2016 → 2017

This setup simulates repeated model development over time while preventing future observations from being used to predict the past.

The folds store only the training and validation indices. The corresponding feature and target data are selected from the development dataset during model evaluation.

## Temporal Baseline Evaluation

Before starting systematic hyperparameter tuning, the previously selected Gradient Boosting configuration (`max_depth=6`) is evaluated across the temporal folds.

The objective of this step is to establish a temporal baseline and verify that model performance is sufficiently stable across validation periods before comparing additional hyperparameter configurations.

In [18]:
fold_f1_scores = []

for train_indices, val_indices in folds:
    X_train = X_development.iloc[train_indices].copy()
    y_train = y_development.iloc[train_indices]

    X_val = X_development.iloc[val_indices].copy()
    y_val = y_development.iloc[val_indices]

    X_train["mcc_key"] = X_train["mcc_key"].astype(str)
    X_val["mcc_key"] = X_val["mcc_key"].astype(str)

    X_train, X_val = encode_categorical_features(X_train, X_val)
    X_train, X_val = handle_missing_values(X_train, X_val)

    gb = GradientBoostingClassifier(max_depth=6, random_state=42)
    gb.fit(X_train, y_train)

    y_pred = gb.predict(X_val)
    f1 = f1_score(y_val, y_pred)

    fold_f1_scores.append(f1)

fold_f1_scores

[0.7662395050817499,
 0.8132045088566827,
 0.5415053039232194,
 0.8480549199084668,
 0.036231884057971016]

## Temporal Validation Diagnostics

The temporal baseline revealed substantial variation in model performance across validation periods. While the model achieved relatively strong F1-scores in 2013, 2014 and 2016, performance decreased considerably in 2015 and collapsed almost completely in 2017.

Before proceeding with hyperparameter tuning, these periods are investigated to determine whether the instability is associated with changes in class prevalence, transaction characteristics, or fraud patterns over time.

The analysis focuses particularly on two distinct failure modes:

- **2015:** reduced precision and a large increase in false positives.
- **2017:** extremely low recall and failure to identify most fraudulent transactions.

In [19]:
yearly_fraud = pd.DataFrame({
    "year": development_years,
    "is_fraud": y_development
})

fraud_by_year = yearly_fraud.groupby("year")["is_fraud"].agg(
    transactions="count",
    frauds="sum",
    fraud_rate="mean"
)

fraud_by_year["fraud_rate"] *= 100

fraud_by_year

,transactions,frauds,fraud_rate
year,,,
2010,37148,2573,6.926349
2011,36048,37,0.102641
2012,37815,923,2.440830
2013,39124,1337,3.417340
2014,38803,664,1.711208
2015,40896,2189,5.352602
2016,41251,2448,5.934402
2017,39258,172,0.438127


### Model Performance Across Validation Periods

Fraud prevalence varies substantially over time. However, changes in class prevalence alone do not explain the differences in model performance.

To understand how the model fails in each validation period, precision, recall and confusion-matrix components are examined separately.

In [20]:
fold_results = []

validation_years = [2013, 2014, 2015, 2016, 2017]

for (train_indices, val_indices), year in zip(folds, validation_years):
    X_train = X_development.iloc[train_indices].copy()
    y_train = y_development.iloc[train_indices]

    X_val = X_development.iloc[val_indices].copy()
    y_val = y_development.iloc[val_indices]

    X_train["mcc_key"] = X_train["mcc_key"].astype(str)
    X_val["mcc_key"] = X_val["mcc_key"].astype(str)

    X_train, X_val = encode_categorical_features(X_train, X_val)
    X_train, X_val = handle_missing_values(X_train, X_val)

    gb = GradientBoostingClassifier(max_depth=6, random_state=42)
    gb.fit(X_train, y_train)

    y_pred = gb.predict(X_val)

    tn, fp, fn, tp = confusion_matrix(y_val, y_pred).ravel()

    fold_results.append({
        "year": year,
        "precision": precision_score(y_val, y_pred),
        "recall": recall_score(y_val, y_pred),
        "f1": f1_score(y_val, y_pred),
        "tn": tn,
        "fp": fp,
        "fn": fn,
        "tp": tp
    })

fold_results = pd.DataFrame(fold_results)

fold_results

,year,precision,recall,f1,tn,fp,fn,tp
0,2013,0.936285,0.648467,0.766240,37728,59,470,867
1,2014,0.873702,0.760542,0.813205,38066,73,159,505
2,2015,0.428800,0.734582,0.541505,36565,2142,581,1608
3,2016,0.964100,0.756944,0.848055,38734,69,595,1853
4,2017,0.048077,0.029070,0.036232,38987,99,167,5


In [21]:
fold_results["actual_fraud_rate"] = (
    (fold_results["tp"] + fold_results["fn"]) /
    (fold_results["tn"] + fold_results["fp"] + fold_results["fn"] + fold_results["tp"])
) * 100

fold_results["predicted_fraud_rate"] = (
    (fold_results["tp"] + fold_results["fp"]) /
    (fold_results["tn"] + fold_results["fp"] + fold_results["fn"] + fold_results["tp"])
) * 100

fold_results[
    ["year", "actual_fraud_rate", "predicted_fraud_rate"]
]

,year,actual_fraud_rate,predicted_fraud_rate
0,2013,3.417340,2.366834
1,2014,1.711208,1.489576
2,2015,5.352602,9.169601
3,2016,5.934402,4.659281
4,2017,0.438127,0.264914


### Temporal Changes in Transaction Characteristics

The error analysis indicates that the weaker validation periods exhibit different failure modes. Before investigating 2015 and 2017 individually, transaction characteristics are compared across years for both fraudulent and legitimate observations.

The analysis considers:

- transaction amount relative to historical user and card behavior,
- transaction channel (`use_chip`),
- whether observed changes affect fraud, legitimate transactions, or both.

These comparisons provide context for the period-specific diagnostics that follow.

In [22]:
fraud_features = X_development.copy()

fraud_features["year"] = development_years.values
fraud_features["is_fraud"] = y_development.values

fraud_features = fraud_features[fraud_features["is_fraud"] == True]

fraud_feature_medians = fraud_features.groupby("year")[
    [
        "amount",
        "user_previous_avg_amount",
        "amount_vs_user_avg",
        "card_previous_avg_amount",
        "amount_vs_card_avg"
    ]
].median()

fraud_feature_medians

,amount,user_previous_avg_amount,amount_vs_user_avg,card_previous_avg_amount,amount_vs_card_avg
year,,,,,
2010,74.290,43.540,28.830,44.160,29.180
2011,99.890,59.670,40.550,58.000,40.720
2012,47.710,41.000,4.750,38.990,7.010
2013,86.110,41.500,43.900,41.170,47.230
2014,87.495,41.930,40.560,38.795,44.515
2015,88.790,41.070,45.730,39.340,46.800
2016,84.665,40.695,43.540,39.700,44.095
2017,29.495,43.460,-10.925,37.610,-7.270


In [23]:
nonfraud_features = X_development.copy()

nonfraud_features["year"] = development_years.values
nonfraud_features["is_fraud"] = y_development.values

nonfraud_features = nonfraud_features[
    nonfraud_features["is_fraud"] == False
]

nonfraud_feature_medians = nonfraud_features.groupby("year")[
    [
        "amount",
        "user_previous_avg_amount",
        "amount_vs_user_avg",
        "card_previous_avg_amount",
        "amount_vs_card_avg"
    ]
].median()

nonfraud_feature_medians

,amount,user_previous_avg_amount,amount_vs_user_avg,card_previous_avg_amount,amount_vs_card_avg
year,,,,,
2010,29.72,40.300,-9.85,39.29,-8.190
2011,29.24,40.330,-10.67,39.44,-8.740
2012,28.97,40.245,-10.43,39.00,-8.520
2013,29.36,40.440,-10.53,39.19,-8.680
2014,28.98,40.150,-10.56,38.70,-8.450
2015,28.84,39.960,-10.73,38.64,-8.770
2016,28.93,40.000,-10.85,38.72,-9.025
2017,28.42,39.880,-10.87,38.43,-8.950


In [24]:
fraud_use_chip_by_year = pd.crosstab(
    fraud_features["year"],
    fraud_features["use_chip"],
    normalize="index"
) * 100

fraud_use_chip_by_year

use_chip,Chip Transaction,Online Transaction,Swipe Transaction
year,,,
2010,0.000000,90.050525,9.949475
2011,0.000000,94.594595,5.405405
2012,0.000000,75.081257,24.918743
2013,0.000000,84.293194,15.706806
2014,0.000000,84.638554,15.361446
2015,8.725445,85.153038,6.121517
2016,7.761438,85.620915,6.617647
2017,98.837209,0.000000,1.162791


In [25]:
nonfraud_use_chip_by_year = pd.crosstab(
    nonfraud_features["year"],
    nonfraud_features["use_chip"],
    normalize="index"
) * 100

nonfraud_use_chip_by_year

use_chip,Chip Transaction,Online Transaction,Swipe Transaction
year,,,
2010,0.000000,10.672451,89.327549
2011,0.000000,10.827247,89.172753
2012,0.000000,11.034913,88.965087
2013,0.000000,11.490724,88.509276
2014,0.000000,11.610163,88.389837
2015,70.651303,12.318185,17.030511
2016,71.115635,12.140814,16.743551
2017,70.861690,12.127104,17.011206


### 2015 Performance Degradation

The 2015 validation period exhibits a primarily false-positive-driven failure mode.

Although the model retains relatively strong fraud detection ability (**recall = 0.734**), precision decreases to **0.438**, resulting in an F1-score of **0.549**. The model incorrectly classifies **2,057 legitimate transactions as fraud**.

The following analysis investigates whether these false positives are associated with temporal changes in the transaction environment.

#### Recreating the 2015 Validation Fold

The 2015 validation fold is recreated using the same preprocessing steps and Gradient Boosting configuration as in the temporal baseline evaluation. This allows individual prediction errors to be inspected directly.

In [26]:
train_indices, val_indices = folds[2]

X_train_2015 = X_development.iloc[train_indices].copy()
y_train_2015 = y_development.iloc[train_indices]

X_val_2015 = X_development.iloc[val_indices].copy()
y_val_2015 = y_development.iloc[val_indices]

X_train_2015["mcc_key"] = X_train_2015["mcc_key"].astype(str)
X_val_2015["mcc_key"] = X_val_2015["mcc_key"].astype(str)

X_train_2015_encoded, X_val_2015_encoded = encode_categorical_features(
    X_train_2015,
    X_val_2015
)

X_train_2015_encoded, X_val_2015_encoded = handle_missing_values(
    X_train_2015_encoded,
    X_val_2015_encoded
)

gb_2015 = GradientBoostingClassifier(max_depth=6, random_state=42)
gb_2015.fit(X_train_2015_encoded, y_train_2015)

y_pred_2015 = gb_2015.predict(X_val_2015_encoded)

#### False Positive Analysis

Because the 2015 degradation is primarily driven by reduced precision, the analysis focuses on legitimate transactions incorrectly classified as fraud.

First, the transaction-channel composition of the false positives is examined.

In [27]:
false_positive_mask_2015 = (
    (y_val_2015.to_numpy() == False) &
    (y_pred_2015 == True)
)

false_positives_2015 = X_val_2015.loc[false_positive_mask_2015]

false_positives_2015["use_chip"].value_counts(normalize=True) * 100

use_chip
Chip Transaction      97.852474
Online Transaction     2.007470
Swipe Transaction      0.140056
Name: proportion, dtype: float64

Nearly all false positives are Chip Transactions. To determine whether this reflects only the composition of the false-positive group or a broader model problem with legitimate chip transactions, the false-positive rate within that group is calculated next.

In [28]:
nonfraud_chip_mask_2015 = (
    (y_val_2015.to_numpy() == False) &
    (X_val_2015["use_chip"].to_numpy() == "Chip Transaction")
)

nonfraud_chip_predictions_2015 = y_pred_2015[nonfraud_chip_mask_2015]

nonfraud_chip_predictions_2015.mean() * 100

np.float64(7.664460452700479)

#### 2015 Diagnostic Conclusion

The 2015 validation period shows a different failure mode from the severe performance collapse observed in 2017.

Although the model maintained relatively strong fraud detection ability in 2015 (**recall = 0.734**), precision decreased substantially to **0.438**, resulting in an F1-score of **0.549**. The primary source of this degradation was a large increase in false positives: **2,057 legitimate transactions were incorrectly classified as fraud**.

Further analysis revealed a major temporal change in the transaction channel distribution. Until 2014, legitimate transactions contained virtually no `Chip Transaction` observations. In 2015, however, **70.65% of legitimate transactions were chip-based**.

The model, trained only on earlier periods, therefore encountered a transaction channel that was largely absent from its historical training data.

This shift was strongly associated with the observed false-positive problem:

- **97.86% of all false positives in 2015 were Chip Transactions**
- **7.36% of legitimate Chip Transactions were incorrectly classified as fraud**
- fraud recall nevertheless remained relatively high at **0.734**

Interestingly, the distribution of legitimate transaction channels remained almost unchanged in 2016, while model performance recovered to **F1 = 0.848**. Under the expanding-window validation scheme, the 2016 training set already included 2015 observations, giving the model exposure to the newly prevalent chip-based transaction environment.

These results suggest that the 2015 performance degradation is primarily associated with a **temporal shift in the transaction environment and an elevated false-positive rate**, rather than a failure to recognize fraudulent transactions themselves.

The evidence demonstrates a strong association rather than establishing that the channel shift alone caused the performance degradation.

### 2017 Performance Collapse

Unlike 2015, the 2017 validation period is characterized by an almost complete failure to identify fraudulent transactions.

The model achieved:

- **precision = 0.048**
- **recall = 0.029**
- **F1 = 0.036**

Only **5 of 172 fraudulent transactions** were correctly identified.

The following analysis investigates whether this failure is associated with a change in the characteristics of fraudulent transactions in 2017.

#### Recreating the 2017 Validation Fold

The 2017 validation fold is recreated using the same preprocessing steps and Gradient Boosting configuration as in the temporal baseline evaluation. This allows the fraudulent transactions missed by the model to be inspected directly.

In [29]:
train_indices, val_indices = folds[4]

X_train_2017 = X_development.iloc[train_indices].copy()
y_train_2017 = y_development.iloc[train_indices]

X_val_2017 = X_development.iloc[val_indices].copy()
y_val_2017 = y_development.iloc[val_indices]

X_train_2017["mcc_key"] = X_train_2017["mcc_key"].astype(str)
X_val_2017["mcc_key"] = X_val_2017["mcc_key"].astype(str)

X_train_2017_encoded, X_val_2017_encoded = encode_categorical_features(
    X_train_2017,
    X_val_2017
)

X_train_2017_encoded, X_val_2017_encoded = handle_missing_values(
    X_train_2017_encoded,
    X_val_2017_encoded
)

gb_2017 = GradientBoostingClassifier(max_depth=6, random_state=42)
gb_2017.fit(X_train_2017_encoded, y_train_2017)

y_pred_2017 = gb_2017.predict(X_val_2017_encoded)

#### False Negative Analysis

The 2017 performance collapse is driven by extremely low recall. Therefore, the analysis focuses on fraudulent transactions incorrectly classified as legitimate.

The transaction-channel distribution of these false negatives is examined first.

In [30]:
false_negative_mask_2017 = (
    (y_val_2017.to_numpy() == True) &
    (y_pred_2017 == False)
)

false_negative_2017 = X_val_2017.loc[false_negative_mask_2017]

false_negative_2017["use_chip"].value_counts(normalize=True) * 100

use_chip
Chip Transaction     98.802395
Swipe Transaction     1.197605
Name: proportion, dtype: float64

Almost all missed frauds are Chip Transactions. However, because Chip Transactions also account for nearly all fraud in 2017, this alone does not demonstrate that the model specifically fails on this channel.

The detection rate among all fraudulent Chip Transactions is therefore examined next.

In [31]:
fraud_chip_mask_2017 = (
    (y_val_2017.to_numpy() == True) &
    (X_val_2017["use_chip"].to_numpy() == "Chip Transaction")
)

fraud_chip_predictions_2017 = y_pred_2017[fraud_chip_mask_2017]

fraud_chip_predictions_2017.mean() * 100

np.float64(2.941176470588235)

#### Historical vs. 2017 Chip Fraud

The low detection rate for Chip Transactions does not by itself explain the 2017 collapse, because chip-based fraud was also present in earlier training data.

To determine whether the nature of chip-based fraud changed, historical Chip Transactions classified as fraud are separated from Chip Transactions classified as fraud in 2017.

These groups will be compared to determine whether the 2017 fraud pattern differs from the historical chip-fraud observations available to the model.

In [32]:
historical_chip_fraud_mask = (
    (y_development == True) &
    (development_years < 2017) &
    (X_development["use_chip"] == "Chip Transaction")
)

historical_chip_frauds = X_development[historical_chip_fraud_mask]

len(historical_chip_frauds)

381

In [33]:
chip_fraud_2017_mask = (
    (y_development == True) &
    (development_years == 2017) &
    (X_development["use_chip"] == "Chip Transaction")
)

chip_fraud_2017 = X_development[chip_fraud_2017_mask]

len(chip_fraud_2017)

170

In [34]:
historical_chip_frauds_medians = historical_chip_frauds[
    [
        "amount",
        "amount_vs_user_avg",
        "amount_vs_card_avg",
        "user_previous_transaction_count",
        "user_previous_avg_amount",
        "card_previous_transaction_count",
        "card_previous_avg_amount"
    ]
].median()

historical_chip_frauds_medians

amount                               76.17
amount_vs_user_avg                   31.77
amount_vs_card_avg                   34.59
user_previous_transaction_count    6016.00
user_previous_avg_amount             41.05
card_previous_transaction_count    1742.00
card_previous_avg_amount             41.87
dtype: float64

In [35]:
chip_fraud_2017_medians = chip_fraud_2017[
    [
        "amount",
        "amount_vs_user_avg",
        "amount_vs_card_avg",
        "user_previous_transaction_count",
        "user_previous_avg_amount",
        "card_previous_transaction_count",
        "card_previous_avg_amount"
    ]
].median()

chip_fraud_2017_medians

amount                               28.440
amount_vs_user_avg                  -12.015
amount_vs_card_avg                   -7.680
user_previous_transaction_count    7805.000
user_previous_avg_amount             43.455
card_previous_transaction_count    2143.000
card_previous_avg_amount             37.600
dtype: float64

In [36]:
chip_nonfraud_2017_mask = (
    (y_development == False) &
    (development_years == 2017) &
    (X_development["use_chip"] == "Chip Transaction")
)

chip_nonfraud_2017 = X_development[chip_nonfraud_2017_mask]

len(chip_nonfraud_2017)

27697

In [37]:
chip_nonfraud_2017_medians = chip_nonfraud_2017[
    [
        "amount",
        "amount_vs_user_avg",
        "amount_vs_card_avg",
        "user_previous_transaction_count",
        "user_previous_avg_amount",
        "card_previous_transaction_count",
        "card_previous_avg_amount"
    ]
].median()

chip_nonfraud_2017_medians

amount                               26.38
amount_vs_user_avg                  -12.47
amount_vs_card_avg                  -10.16
user_previous_transaction_count    8939.00
user_previous_avg_amount             40.02
card_previous_transaction_count    3130.00
card_previous_avg_amount             38.45
dtype: float64

#### Categorical Feature Comparison

The numerical comparison indicates that fraudulent and legitimate Chip Transactions in 2017 are highly similar across several amount-based behavioral features, particularly `amount_vs_user_avg` and `amount_vs_card_avg`.

To determine whether the two groups can still be distinguished through other transaction characteristics, the analysis is extended to categorical and discrete features.

The following features are examined:

- `mcc_key`
- `merchant_id`
- `transaction_hour`
- `day_of_week`
- `is_weekend`

`use_chip` is excluded because both groups are intentionally restricted to Chip Transactions.

In [38]:
chip_fraud_2017["mcc_key"].value_counts(normalize=True).head(10) * 100

mcc_key
10    19.411765
12    13.529412
11     7.058824
4      5.882353
14     4.705882
35     4.705882
67     3.529412
68     3.529412
25     3.529412
2      3.529412
Name: proportion, dtype: float64

In [39]:
chip_nonfraud_2017["mcc_key"].value_counts(normalize=True).head(10) * 100

mcc_key
4     13.907643
14    12.470665
2     11.957974
1      8.434126
25     6.657761
16     5.137741
9      4.942774
8      4.314547
24     3.953497
10     2.920894
Name: proportion, dtype: float64

In [40]:
historical_chip_frauds["mcc_key"].value_counts(normalize=True).head(10) * 100

mcc_key
16    26.771654
4     12.073491
12     6.036745
19     6.036745
25     4.724409
29     4.461942
35     4.199475
67     3.412073
15     3.412073
3      3.149606
Name: proportion, dtype: float64

In [41]:
(historical_chip_frauds["mcc_key"] == 10).mean() * 100

np.float64(0.0)

In [42]:
mcc10_fraud_2017_mask = (
    (y_val_2017.to_numpy() == True) &
    (X_val_2017["mcc_key"].to_numpy() == "10")
)

mcc10_fraud_2017 = y_pred_2017[mcc10_fraud_2017_mask]

len(mcc10_fraud_2017)

34

In [43]:
mcc10_fraud_2017.mean() * 100

np.float64(0.0)

##### MCC Distribution Shift

The MCC distribution of Chip Transaction frauds changed substantially in 2017.

Most notably, `mcc_key = 10` accounted for approximately 19% of Chip Transaction frauds in 2017, while no historical Chip Transaction frauds belonged to this category.

The model failed to identify any of the 34 MCC 10 fraud cases present in the 2017 validation fold, resulting in a 0% recall for this subgroup.

This suggests that the poor 2017 performance is associated not only with changes in transaction amounts, but also with a shift in the categorical composition of fraudulent transactions. However, this result does not establish MCC 10 as the direct cause of the prediction failures, since other feature changes may occur simultaneously.

In [44]:
chip_fraud_2017["merchant_id"].value_counts()

merchant_id
83018    23
48919    11
99370    11
67570     7
86410     7
53492     6
46284     6
41375     6
80114     5
43293     5
2297      5
59474     5
2156      5
75034     5
31893     4
19463     4
88260     4
22204     4
32070     3
51300     3
4439      3
90915     3
7777      3
16790     3
1744      2
66052     2
66615     2
9391      2
68751     2
83271     2
11468     2
61195     2
59935     2
33191     2
49637     1
24504     1
14528     1
5059      1
45992     1
94625     1
34490     1
17587     1
4906      1
Name: count, dtype: int64

In [45]:
chip_fraud_2017["merchant_id"].value_counts().value_counts().sort_index()

count
1      9
2     10
3      6
4      4
5      6
6      3
7      2
11     2
23     1
Name: count, dtype: int64

In [46]:
historical_chip_fraud_merchants = historical_chip_frauds["merchant_id"].unique()
historical_chip_fraud_merchants

array([ 70082,  76134,  44886,  91727,   5260,  70364,  20303,  44168,
        79447,  18422,  19557,  89216,  51922,  91479,   2504,  12170,
        23192,  16443,   2353,  21952,  91103,  40080,  17530,  27405,
        18987,  71663,  51350,  97517,  77786,  61672,   3739,  28275,
        64903,  63608,  13730,  94963,   4076,  14479,  97077,  78595,
        18786,   2620,  80817,  59438,  78711,    256,  73829,  97909,
         7882,  66080,   9156,  48136,  70625,  90116,  25206,  71196,
        30971,  82353,  86313,  42629,  68947,  26496,  11623,  73578,
        63058,   9178,  84316,  47786,  41548,  72294,  32868,   8635,
        51623,  31947,  49773,  61714,  76245,  40815,  60614,  69807,
        13759,  83316,  12031,  57422,  89285,  64635,  27442,  61904,
        54658,  52573,  30068,   9296,  61530,  39582,  13938,  79466,
        43353,  85820,  88985,  54106,  29153,  19115,  92290,  11593,
        72650,  40054,   5360,  15457,  82456,  64755,  47636,  11901,
      

In [47]:
known_merchant_2017_mask = chip_fraud_2017["merchant_id"].isin(historical_chip_fraud_merchants)
known_merchant_2017_mask.mean() * 100

np.float64(0.0)

In [48]:
historical_merchants = X_train_2017["merchant_id"].unique()
historical_merchants

array([89069, 99370, 39105, ..., 25662, 65172, 38238], shape=(17635,))

In [49]:
known_historical_merchant_2017_mask = chip_fraud_2017["merchant_id"].isin(historical_merchants)
known_historical_merchant_2017_mask.mean() * 100

np.float64(85.88235294117646)

In [50]:
chip_fraud_2017_merchants = chip_fraud_2017["merchant_id"].unique()
len(chip_fraud_2017_merchants)

43

In [51]:
known_unique_merchants_2017_mask = np.isin(
    chip_fraud_2017_merchants,
    historical_merchants
)
known_unique_merchants_2017_mask.mean() * 100

np.float64(79.06976744186046)

In [52]:
known_merchant_chip_fraud_2017_mask = (
    (y_val_2017.to_numpy() == True) &
    (X_val_2017["use_chip"].to_numpy() == "Chip Transaction") &
    (X_val_2017["merchant_id"].isin(historical_merchants).to_numpy())
)
known_merchant_chip_fraud_2017_mask.sum()

np.int64(146)

In [53]:
known_merchant_chip_fraud_predictions_2017 = y_pred_2017[
    known_merchant_chip_fraud_2017_mask
]
known_merchant_chip_fraud_predictions_2017.mean() * 100

np.float64(0.0)

In [54]:
new_merchant_chip_fraud_2017_mask = (
    (y_val_2017.to_numpy() == True) &
    (X_val_2017["use_chip"].to_numpy() == "Chip Transaction") &
    (~X_val_2017["merchant_id"].isin(historical_merchants).to_numpy())
)

new_merchant_chip_fraud_2017_mask.sum()

np.int64(24)

In [55]:
new_merchant_chip_fraud_predictions_2017 = y_pred_2017[
    new_merchant_chip_fraud_2017_mask
]

new_merchant_chip_fraud_predictions_2017.mean() * 100

np.float64(20.833333333333336)

### Fold 5 Diagnostics Conclusion

The exceptionally low F1-score in the 2017 validation fold is associated with a substantial shift in the transaction and fraud patterns compared with the historical training data.

The most important change is the transition towards Chip Transactions. While Chip Transactions were absent before 2015, they account for almost all fraudulent transactions in 2017.

Further analysis indicates that the characteristics of these frauds also changed. Fraudulent Chip Transactions in 2017 resemble legitimate Chip Transactions more closely in terms of transaction amount and user/card spending behaviour than historical Chip frauds.

Categorical distributions changed as well. New fraud patterns appeared within MCC categories, including MCC 10, which was absent from historical Chip frauds and achieved 0% recall in the 2017 validation fold.

Merchant analysis provides additional evidence that the issue is not simply caused by previously unseen merchants. Most 2017 Chip frauds occurred at merchants already present in the training data, yet the model failed to detect these cases.

Overall, the results suggest temporal concept drift between the historical training data and the 2017 fraud patterns. This explains why the final temporal validation fold performs substantially worse than the earlier folds and highlights the importance of evaluating fraud detection models using time-aware validation.

## Gradient Boosting Hyperparameter Tuning

The temporal cross-validation experiments revealed substantial variation in model performance across validation years, including a pronounced deterioration in the final 2017 fold associated with temporal distribution shift.

Hyperparameter tuning will now be performed across the same temporal folds within the 2010–2017 development period. Each Gradient Boosting configuration will therefore be evaluated not only on its average F1-score, but also on its performance stability across time.

The objective is to determine whether hyperparameter optimization can improve the model's overall temporal robustness. Importantly, tuning cannot introduce information about future fraud patterns that is absent from the historical training data, so the 2017 performance will also indicate the limitations of standard model optimization under concept drift.

The selected Gradient Boosting configuration will subsequently be evaluated on the untouched 2018 validation period.

In [56]:
max_depths = [2, 6, 10]
all_depths_scores = []


for i in max_depths:
    fold_f1_scores = []
    
    for train_indices, val_indices in folds:
        X_train = X_development.iloc[train_indices].copy()
        y_train = y_development.iloc[train_indices]
    
        X_val = X_development.iloc[val_indices].copy()
        y_val = y_development.iloc[val_indices]
    
        X_train["mcc_key"] = X_train["mcc_key"].astype(str)
        X_val["mcc_key"] = X_val["mcc_key"].astype(str)
    
        X_train, X_val = encode_categorical_features(X_train, X_val)
        X_train, X_val = handle_missing_values(X_train, X_val)
    
        gb = GradientBoostingClassifier(max_depth=i, random_state=42)
        gb.fit(X_train, y_train)
    
        y_pred = gb.predict(X_val)
        f1 = f1_score(y_val, y_pred)
    
        fold_f1_scores.append(f1)
    all_depths_scores.append(fold_f1_scores)

all_depths_scores

[[0.23958333333333334,
  0.5288562434417629,
  0.49307666328942923,
  0.6396103896103896,
  0.0],
 [0.7662395050817499,
  0.8132045088566827,
  0.5415053039232194,
  0.8480549199084668,
  0.036231884057971016],
 [0.8036835496023441,
  0.8351309707241911,
  0.4919011082693947,
  0.8604444444444445,
  0.0763888888888889]]

In [57]:
for max_depth, scores in zip(max_depths, all_depths_scores):
    print(f"max_depth={max_depth}: mean F1 = {np.mean(scores):.4f}")

max_depth=2: mean F1 = 0.3802
max_depth=6: mean F1 = 0.6010
max_depth=10: mean F1 = 0.6135


### Initial `max_depth` Evaluation

As an initial step, three substantially different tree depths were evaluated using temporal cross-validation.

Increasing `max_depth` from 2 to 6 resulted in a substantial improvement in mean F1-score, while the improvement from 6 to 10 was considerably smaller. This confirms the earlier observation that tree depth has a strong influence on Gradient Boosting performance, while also suggesting diminishing returns as model complexity increases.

Rather than continuing to test individual depth values manually, the next step is to evaluate multiple Gradient Boosting hyperparameters systematically using the same temporal cross-validation framework.

In [62]:
param_distributions = {
    "model__max_depth": [4, 6, 8, 10, 12, 15],
    "model__learning_rate": [0.01, 0.05, 0.1, 0.2],
    "model__n_estimators": [100, 250, 500, 1000]
}

In [63]:
categorical_columns = list(
    X_development.select_dtypes(include=["object", "string"]).columns
)

if "mcc_key" not in categorical_columns:
    categorical_columns.append("mcc_key")

numerical_columns = [
    col for col in X_development.columns
    if col not in categorical_columns
]
print("Categorical:", categorical_columns)
print("Numerical:", numerical_columns)

Categorical: ['use_chip', 'mcc_key']
Numerical: ['amount', 'merchant_id', 'transaction_hour', 'day_of_week', 'is_weekend', 'user_previous_transaction_count', 'user_previous_avg_amount', 'amount_vs_user_avg', 'card_previous_transaction_count', 'card_previous_avg_amount', 'amount_vs_card_avg']


In [64]:
preprocessor = ColumnTransformer(
    transformers=[
        (
            "categorical",
            OneHotEncoder(
                handle_unknown="ignore",
                sparse_output=False
            ),
            categorical_columns
        ),
        (
            "numerical",
            SimpleImputer(strategy="median"),
            numerical_columns
        )
    ]
)

In [65]:
pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", GradientBoostingClassifier(random_state=42))
])

In [66]:
rs = RandomizedSearchCV(
    estimator=pipeline,
    param_distributions=param_distributions,
    n_iter=20,
    scoring="f1",
    cv=folds,
    random_state=42
)

In [67]:
train_indices, val_indices = folds[0]

X_train_test = X_development.iloc[train_indices]
y_train_test = y_development.iloc[train_indices]

X_val_test = X_development.iloc[val_indices]

pipeline.fit(X_train_test, y_train_test)

y_pred_test = pipeline.predict(X_val_test)

print("Pipeline works correctly.")
print("Predictions:", len(y_pred_test))

Pipeline works correctly.
Predictions: 39124


In [68]:
rs.fit(X_development, y_development)

KeyboardInterrupt: 